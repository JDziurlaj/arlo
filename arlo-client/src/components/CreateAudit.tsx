import React, { useState, useEffect, useCallback } from 'react'
import styled from 'styled-components'
import moment from 'moment'
import { toast } from 'react-toastify'
import { RouteComponentProps, Link } from 'react-router-dom'
import { RadioGroup, Radio } from '@blueprintjs/core'
import FormButton from './Form/FormButton'
import { api, checkAndToast } from './utilities'
import { ICreateAuditParams, IErrorResponse } from '../types'

const Button = styled(FormButton)`
  margin: 65px 0;
`

const Wrapper = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  text-align: center;
  @media (min-width: 500px) {
    width: 400px;
  }
`

interface IElection {
  id: string
  name: string
  date: string
}
export interface IElections {
  elections: IElection[]
}

const CreateAudit = ({ history }: RouteComponentProps<ICreateAuditParams>) => {
  const [loading, setLoading] = useState(false)
  const [electionId, setElectionId] = useState('')

  const onClick = async () => {
    try {
      setLoading(true)
      const response: { electionId: string } | IErrorResponse = await api(
        '/election/new',
        {
          method: 'POST',
        }
      )
      if (checkAndToast(response)) {
        return
      }
      const { electionId } = response
      history.push(`/election/${electionId}`)
    } catch (err) {
      toast.error(err.message)
    }
  }

  const [elections, setElections] = useState<IElections>({ elections: [] })

  const getStatus = useCallback(async (): Promise<IElections> => {
    try {
      const list: IElections | IErrorResponse = await api(`/elections`)
      if (checkAndToast(list)) {
        return { elections: [] }
      } else {
        return list
      }
    } catch (err) {
      toast.error(err.message)
    }
    return { elections: [] }
  }, [])

  const updateElections = useCallback(async () => {
    const list = await getStatus()
    setLoading(true)
    setElections(list)
    setElectionId(list.elections.length ? list.elections[0].id : '')
    setLoading(false)
  }, [getStatus])

  useEffect(() => {
    updateElections()
  }, [updateElections])

  return (
    <Wrapper>
      <img height="50px" src="/arlo.png" alt="Arlo, by VotingWorks" />
      <Button
        type="button"
        intent="primary"
        fill
        large
        onClick={onClick}
        loading={loading}
        disabled={loading}
      >
        Create a New Audit
      </Button>
      {elections.elections.length > 0 ? (
        <>
          <RadioGroup
            label="Select an existing audit:"
            selectedValue={electionId}
            onChange={e => setElectionId(e.currentTarget.value)}
          >
            {elections.elections.map(e => (
              <Radio key={e.id} value={e.id}>
                {e.name || <i>Not named yet</i>} (Created&nbsp;
                {moment(e.date).format('MM/DD/YYYY')})
              </Radio>
            ))}
          </RadioGroup>
          <Link
            to={`/election/${electionId}`}
            className="bp3-button bp3-intent-primary"
          >
            View Audit
          </Link>
        </>
      ) : (
        <p>
          You do not have any audits currently associated with your account.
        </p>
      )}
    </Wrapper>
  )
}

export default CreateAudit
