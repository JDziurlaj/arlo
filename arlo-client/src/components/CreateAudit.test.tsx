import React from 'react'
import { render, fireEvent, wait } from '@testing-library/react'
import { toast } from 'react-toastify'
import { RouteComponentProps, BrowserRouter as Router } from 'react-router-dom'
import CreateAudit, { IElections } from './CreateAudit'
import { ICreateAuditParams } from '../types'
import { routerTestProps } from './testUtilities'
import * as utilities from './utilities'

const apiMock: jest.SpyInstance<
  ReturnType<typeof utilities.api>,
  Parameters<typeof utilities.api>
> = jest.spyOn(utilities, 'api').mockImplementation(
  async (
    endpoint: string
  ): Promise<IElections | { electionId: string } | undefined> => {
    switch (endpoint) {
      case '/election/new':
        return { electionId: '1' }
      case '/elections':
        return {
          elections: [
            {
              name: 'Election One',
              id: 'election-1',
              date: 'Thu, 18 Jul 2019 16:34:07 GMT',
            },
          ],
        }
      default:
        return undefined
    }
  }
)
const checkAndToastMock: jest.SpyInstance<
  ReturnType<typeof utilities.checkAndToast>,
  Parameters<typeof utilities.checkAndToast>
> = jest.spyOn(utilities, 'checkAndToast').mockReturnValue(false)

checkAndToastMock.mockReturnValue(false)

const routeProps: RouteComponentProps<ICreateAuditParams> = routerTestProps(
  '/election/:electionId',
  {
    electionId: '1',
  }
)

const toastSpy = jest.spyOn(toast, 'error').mockImplementation()

const historySpy = jest.spyOn(routeProps.history, 'push').mockImplementation()

afterEach(() => {
  apiMock.mockClear()
  checkAndToastMock.mockClear()
  toastSpy.mockClear()
  historySpy.mockClear()
})

describe('CreateAudit', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Router>
        <CreateAudit {...routeProps} />
      </Router>
    )
    expect(container).toMatchSnapshot()
  })

  it('calls the /election/new endpoint', async () => {
    const { getByText } = render(
      <Router>
        <CreateAudit {...routeProps} />
      </Router>
    )

    fireEvent.click(getByText('Create a New Audit'), { bubbles: true })

    await wait(() => {
      expect(apiMock).toBeCalledTimes(2)
      expect(apiMock.mock.calls[0][0]).toBe('/elections')
      expect(apiMock.mock.calls[1][0]).toBe('/election/new')
      expect(historySpy).toBeCalledTimes(1)
      expect(historySpy.mock.calls[0][0]).toBe('/election/1')
    })
  })

  it('handles error responses from server', async () => {
    checkAndToastMock.mockReturnValue(true)
    const { getByText } = render(
      <Router>
        <CreateAudit {...routeProps} />
      </Router>
    )

    fireEvent.click(getByText('Create a New Audit'), { bubbles: true })

    await wait(() => {
      expect(apiMock).toBeCalledTimes(2)
      expect(apiMock.mock.calls[0][0]).toBe('/elections')
      expect(apiMock.mock.calls[1][0]).toBe('/election/new')
      expect(checkAndToastMock).toBeCalledTimes(2)
      expect(historySpy).toBeCalledTimes(0)
      expect(toastSpy).toBeCalledTimes(0)
    })
  })

  it('handles 404 responses from server', async () => {
    apiMock.mockImplementation(async () => {
      throw new Error('404')
    })
    checkAndToastMock.mockReturnValue(true)
    const { getByText } = render(
      <Router>
        <CreateAudit {...routeProps} />
      </Router>
    )

    fireEvent.click(getByText('Create a New Audit'), { bubbles: true })

    await wait(() => {
      expect(apiMock).toBeCalledTimes(2)
      expect(apiMock.mock.calls[0][0]).toBe('/elections')
      expect(apiMock.mock.calls[1][0]).toBe('/election/new')
      expect(checkAndToastMock).toBeCalledTimes(0)
      expect(historySpy).toBeCalledTimes(0)
      expect(toastSpy).toBeCalledTimes(2)
    })
  })
})
