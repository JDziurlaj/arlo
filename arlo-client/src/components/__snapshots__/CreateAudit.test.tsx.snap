// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAudit renders correctly 1`] = `
<div>
  <div
    class="sc-htpNat dgtpPg"
  >
    <img
      alt="Arlo, by VotingWorks"
      height="50px"
      src="/arlo.png"
    />
    <button
      class="bp3-button bp3-fill bp3-large bp3-intent-primary sc-bwzfXH iEhHA sc-bdVaJa gLVjYw"
      type="button"
    >
      <span
        class="bp3-button-text"
      >
        Create a New Audit
      </span>
    </button>
    <div>
      <label
        class="bp3-label"
      >
        Select an existing audit:
      </label>
    </div>
    <a
      class="bp3-button bp3-intent-primary"
      href="/election/"
    >
      View Audit
    </a>
  </div>
</div>
`;
